File: SystematicCredit/scripts/build_dashboard.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const __filename = fileURLToPath(import.meta.url);
const PROJECT_ROOT = path.resolve(path.dirname(__filename), "..");
const PAYLOAD_PATH = path.join(PROJECT_ROOT, "data", "output", "dashboard_payload.json");
const OUTPUT_DIR = path.join(PROJECT_ROOT, "outputs", "systematic_credit_toolkit");
const PREVIEW_DIR = path.join(OUTPUT_DIR, "previews");
const OUTPUT_XLSX = path.join(OUTPUT_DIR, "bond_cds_basis_dashboard.xlsx");
const WORKBOOK_ROW_LIMIT = 1200;

const SHEET_ORDER = [
  "Dashboard",
  "CDS_Curves",
  "Hazard_Survival",
  "Bond_CDS_Basis",
  "RV_Signals",
  "Positions",
  "Macro_Regime",
  "Hedge_Overlay",
  "PnL_Attribution",
  "Performance",
  "Stress_Test",
  "Config",
];

const TABLE_TO_SHEET = {
  cds_curves: "CDS_Curves",
  hazard_survival: "Hazard_Survival",
  bond_cds_basis: "Bond_CDS_Basis",
  rv_signals: "RV_Signals",
  positions: "Positions",
  macro: "Macro_Regime",
  hedge_overlay: "Hedge_Overlay",
  pnl_attribution: "PnL_Attribution",
  performance: "Performance",
  stress_test: "Stress_Test",
  config: "Config",
};

const COLORS = {
  navy: "#12355B",
  blue: "#2563EB",
  teal: "#0F766E",
  green: "#15803D",
  amber: "#B45309",
  red: "#B91C1C",
  gray: "#F3F4F6",
  border: "#CBD5E1",
  text: "#111827",
};

await fs.mkdir(OUTPUT_DIR, { recursive: true });
await fs.mkdir(PREVIEW_DIR, { recursive: true });

const payload = JSON.parse(await fs.readFile(PAYLOAD_PATH, "utf8"));
const workbook = Workbook.create();
const sheets = new Map();
const renderRanges = new Map([["Dashboard", "A1:Q104"]]);

for (const sheetName of SHEET_ORDER) {
  const sheet = workbook.worksheets.add(sheetName);
  sheet.showGridLines = false;
  sheets.set(sheetName, sheet);
}

for (const [tableName, sheetName] of Object.entries(TABLE_TO_SHEET)) {
  const table = payload.tables[tableName];
  if (!table) continue;
  const range = writeDataSheet(sheets.get(sheetName), table, tableName);
  renderRanges.set(sheetName, range);
}

writeDashboard(sheets.get("Dashboard"), payload);

const dashboardInspect = await workbook.inspect({
  kind: "table",
  range: "Dashboard!A1:H20",
  include: "values,formulas",
  tableMaxRows: 20,
  tableMaxCols: 8,
});
console.log(dashboardInspect.ndjson);

const errorScan = await workbook.inspect({
  kind: "match",
  searchTerm: "#REF!|#DIV/0!|#VALUE!|#NAME\\?|#N/A",
  options: { useRegex: true, maxResults: 300 },
  summary: "final formula error scan",
});
console.log(errorScan.ndjson);

for (const sheetName of SHEET_ORDER) {
  const preview = await workbook.render({
    sheetName,
    range: renderRanges.get(sheetName),
    scale: 1,
    format: "png",
  });
  const previewBytes = new Uint8Array(await preview.arrayBuffer());
  await fs.writeFile(path.join(PREVIEW_DIR, `${sheetName}.png`), previewBytes);
}

const output = await SpreadsheetFile.exportXlsx(workbook);
await output.save(OUTPUT_XLSX);
console.log(`Dashboard written to ${OUTPUT_XLSX}`);

function writeDataSheet(sheet, table, tableName) {
  const columns = table.columns;
  const records = limitWorkbookRows(table.records, tableName);
  const matrix = [columns.map(formatHeader)];
  for (const record of records) {
    matrix.push(columns.map((column) => toCellValue(record[column], column)));
  }

  const rowCount = matrix.length;
  const colCount = columns.length;
  const range = sheet.getRangeByIndexes(0, 0, rowCount, colCount);
  range.values = matrix;

  const tableRange = `A1:${colToLetter(colCount)}${rowCount}`;
  const excelTable = sheet.tables.add(tableRange, true, toTableName(tableName));
  excelTable.style = "TableStyleMedium2";
  excelTable.showFilterButton = true;

  sheet.freezePanes.freezeRows(1);
  sheet.getRangeByIndexes(0, 0, 1, colCount).format = {
    fill: COLORS.navy,
    font: { bold: true, color: "#FFFFFF" },
  };
  sheet.getRangeByIndexes(0, 0, rowCount, colCount).format.borders = {
    preset: "outside",
    style: "thin",
    color: COLORS.border,
  };

  for (let colIdx = 0; colIdx < colCount; colIdx += 1) {
    const columnName = columns[colIdx];
    const colRange = sheet.getRangeByIndexes(1, colIdx, Math.max(rowCount - 1, 1), 1);
    colRange.format.numberFormat = [[numberFormatFor(columnName)]];
    sheet.getRangeByIndexes(0, colIdx, rowCount, 1).format.columnWidth = widthFor(columnName);
  }
  return `A1:${colToLetter(Math.min(colCount, 12))}${Math.min(rowCount, 30)}`;
}

function limitWorkbookRows(records, tableName) {
  const unlimited = new Set([
    "macro",
    "hedge_overlay",
    "pnl_attribution",
    "performance",
    "stress_test",
    "config",
  ]);
  if (unlimited.has(tableName) || records.length <= WORKBOOK_ROW_LIMIT) {
    return records;
  }
  return records.slice(records.length - WORKBOOK_ROW_LIMIT);
}

function writeDashboard(sheet, payload) {
  sheet.getRange("A1:H1").merge();
  sheet.getRange("A1").values = [["Bond-CDS Basis Trading Dashboard"]];
  sheet.getRange("A1:H1").format = {
    fill: COLORS.navy,
    font: { bold: true, color: "#FFFFFF", size: 16 },
  };
  sheet.getRange("A2:H2").merge();
  sheet.getRange("A2").values = [["Synthetic issuer-level research dashboard | CDS curves, basis signals, CS01 exposure, hedging, and P&L attribution"]];
  sheet.getRange("A2:H2").format = {
    fill: "#E5E7EB",
    font: { color: COLORS.text, italic: true },
  };

  writeKpiBlock(sheet, payload);
  writeDashboardHelpers(sheet, payload);
  addDashboardCharts(sheet);
  formatDashboard(sheet);
}

function writeKpiBlock(sheet, payload) {
  const hedgeRows = payload.tables.hedge_overlay.records.length;
  const hedgeLastRow = hedgeRows + 1;
  const kpiLabels = [
    ["Metric", "Value", "Format"],
    ["Total P&L", null, "Currency"],
    ["Total Return", null, "Percent"],
    ["Sharpe Ratio", null, "Number"],
    ["Max Drawdown", null, "Percent"],
    ["Hit Rate", null, "Percent"],
    ["Ending NAV", null, "Currency"],
  ];
  sheet.getRange("A4:C10").values = kpiLabels;
  sheet.getRange("B5:B10").formulas = [
    ["='Performance'!B2"],
    ["='Performance'!C2"],
    ["='Performance'!E2"],
    ["='Performance'!F2"],
    ["='Performance'!G2"],
    ["='Performance'!H2"],
  ];
  sheet.getRange("E4:G10").values = [
    ["Latest Risk", "Value", "Unit"],
    ["Macro Regime", null, "label"],
    ["Net Strategy CS01", null, "per bp"],
    ["Gross Strategy CS01", null, "per bp"],
    ["Hedge Ratio", null, "%"],
    ["Hedge CS01", null, "per bp"],
    ["Hedge Notional", null, "currency"],
  ];
  sheet.getRange("F5:F10").formulas = [
    [`='Hedge_Overlay'!B${hedgeLastRow}`],
    [`='Hedge_Overlay'!C${hedgeLastRow}`],
    [`='Hedge_Overlay'!D${hedgeLastRow}`],
    [`='Hedge_Overlay'!E${hedgeLastRow}`],
    [`='Hedge_Overlay'!F${hedgeLastRow}`],
    [`='Hedge_Overlay'!G${hedgeLastRow}`],
  ];
}

function writeDashboardHelpers(sheet, payload) {
  const pnlRows = payload.tables.pnl_attribution.records.length;
  const pnlStart = Math.max(0, pnlRows - 60);
  const pnlCount = pnlRows - pnlStart;
  sheet.getRange("A13:B13").values = [["Date", "Cumulative P&L"]];
  const pnlFormulas = [];
  for (let i = 0; i < pnlCount; i += 1) {
    const sourceRow = 2 + pnlStart + i;
    pnlFormulas.push([
      `=TEXT('PnL_Attribution'!A${sourceRow},"yyyy-mm-dd")`,
      `='PnL_Attribution'!H${sourceRow}`,
    ]);
  }
  if (pnlFormulas.length > 0) {
    sheet.getRangeByIndexes(13, 0, pnlFormulas.length, 2).formulas = pnlFormulas;
  }

  const latestBasis = payload.tables.dashboard_summary.records.filter(
    (row) => row.summary_type === "latest_basis",
  );
  sheet.getRange("D13:G13").values = [["Bond", "Basis bps", "Z-score", "Signal"]];
  if (latestBasis.length > 0) {
    sheet.getRangeByIndexes(13, 3, latestBasis.length, 4).values = latestBasis.map((row) => [
      row.bond_id,
      row.bond_cds_basis_bps,
      row.basis_zscore,
      row.rv_signal,
    ]);
  }

  const regimeCounts = payload.tables.dashboard_summary.records.filter(
    (row) => row.summary_type === "regime_counts",
  );
  sheet.getRange("A77:B77").values = [["Regime", "Days"]];
  if (regimeCounts.length > 0) {
    sheet.getRangeByIndexes(77, 0, regimeCounts.length, 2).values = regimeCounts.map((row) => [
      row.macro_regime,
      row.days,
    ]);
  }

  sheet.getRange("D30:E35").values = [
    ["P&L Component", "Total"],
    ["Basis Convergence", null],
    ["Credit Spread", null],
    ["Hedge", null],
    ["Carry", null],
    ["Transaction Costs", null],
  ];
  const pnlLastRow = pnlRows + 1;
  sheet.getRange("E31:E35").formulas = [
    [`=SUM('PnL_Attribution'!B2:B${pnlLastRow})`],
    [`=SUM('PnL_Attribution'!C2:C${pnlLastRow})`],
    [`=SUM('PnL_Attribution'!F2:F${pnlLastRow})`],
    [`=SUM('PnL_Attribution'!D2:D${pnlLastRow})`],
    [`=-SUM('PnL_Attribution'!E2:E${pnlLastRow})`],
  ];
}

function addDashboardCharts(sheet) {
  const pnlChart = sheet.charts.add("line", sheet.getRange("A13:B73"));
  pnlChart.title = "Cumulative P&L";
  pnlChart.hasLegend = false;
  pnlChart.xAxis = { axisType: "textAxis", textStyle: { fontSize: 9 } };
  pnlChart.yAxis = { numberFormatCode: "$#,##0" };
  pnlChart.setPosition("J3", "Q18");

  const basisChart = sheet.charts.add("bar", sheet.getRange("D13:E25"));
  basisChart.title = "Latest Bond-CDS Basis (bps)";
  basisChart.hasLegend = false;
  basisChart.xAxis = { textStyle: { fontSize: 8 } };
  basisChart.yAxis = { numberFormatCode: "#,##0.0" };
  basisChart.setPosition("J20", "Q35");

  const attributionChart = sheet.charts.add("bar", sheet.getRange("D30:E35"));
  attributionChart.title = "P&L Attribution";
  attributionChart.hasLegend = false;
  attributionChart.yAxis = { numberFormatCode: "$#,##0" };
  attributionChart.setPosition("J37", "Q52");

  const regimeChart = sheet.charts.add("bar", sheet.getRange("A77:B82"));
  regimeChart.title = "Macro Regime Days";
  regimeChart.hasLegend = false;
  regimeChart.yAxis = { numberFormatCode: "#,##0" };
  regimeChart.setPosition("A88", "H104");
}

function formatDashboard(sheet) {
  sheet.getRange("A1:Q104").format.font = { name: "Aptos", color: COLORS.text };
  sheet.getRange("A4:C4").format = headerFormat();
  sheet.getRange("E4:G4").format = headerFormat();
  sheet.getRange("A13:B13").format = headerFormat();
  sheet.getRange("D13:G13").format = headerFormat();
  sheet.getRange("A77:B77").format = headerFormat();
  sheet.getRange("D30:E30").format = headerFormat();
  sheet.getRange("B5:B5").format.numberFormat = "$#,##0;[Red]($#,##0);-";
  sheet.getRange("B6:B6").format.numberFormat = "0.0%;[Red](0.0%);-";
  sheet.getRange("B7:B7").format.numberFormat = "0.00";
  sheet.getRange("B8:B9").format.numberFormat = "0.0%;[Red](0.0%);-";
  sheet.getRange("B10:B10").format.numberFormat = "$#,##0;[Red]($#,##0);-";
  sheet.getRange("A14:A73").format.numberFormat = "yyyy-mm-dd";
  sheet.getRange("F6:F7").format.numberFormat = "#,##0";
  sheet.getRange("F8:F8").format.numberFormat = "0.0%";
  sheet.getRange("F9:F9").format.numberFormat = "#,##0";
  sheet.getRange("F10:F10").format.numberFormat = "$#,##0;[Red]($#,##0);-";
  sheet.getRange("B14:B73").format.numberFormat = "$#,##0;[Red]($#,##0);-";
  sheet.getRange("E14:F25").format.numberFormat = "#,##0.0";
  sheet.getRange("E31:E35").format.numberFormat = "$#,##0;[Red]($#,##0);-";
  sheet.getRange("A1:Q104").format.borders = { preset: "outside", style: "thin", color: COLORS.border };
  sheet.getRange("A1:H1").format = {
    fill: COLORS.navy,
    font: { bold: true, color: "#FFFFFF", size: 16, name: "Aptos" },
  };
  sheet.getRange("A2:H2").format = {
    fill: "#E5E7EB",
    font: { color: COLORS.text, italic: true, name: "Aptos" },
  };
  for (let colIdx = 0; colIdx < 17; colIdx += 1) {
    sheet.getRangeByIndexes(0, colIdx, 104, 1).format.columnWidth = colIdx < 8 ? 15 : 12;
  }
}

function headerFormat() {
  return {
    fill: COLORS.navy,
    font: { bold: true, color: "#FFFFFF" },
  };
}

function formatHeader(value) {
  return String(value)
    .replaceAll("_", " ")
    .replace(/\b\w/g, (char) => char.toUpperCase());
}

function toCellValue(value, columnName) {
  if (value === undefined || value === null) return null;
  if (columnName.toLowerCase().includes("date") && typeof value === "string") {
    return new Date(`${value}T00:00:00`);
  }
  return value;
}

function numberFormatFor(columnName) {
  const col = columnName.toLowerCase();
  if (col.includes("date")) return "yyyy-mm-dd";
  if (col.includes("return") || col.includes("probability") || col.includes("hazard_rate") || col.includes("ratio") || col.includes("drawdown") || col.includes("hit_rate")) {
    return "0.0%;[Red](0.0%);-";
  }
  if (col.includes("pnl") || col.includes("notional") || col.includes("cost") || col.includes("nav")) {
    return "$#,##0;[Red]($#,##0);-";
  }
  if (col.includes("bps") || col.includes("cs01") || col.includes("spread") || col.includes("basis")) {
    return "#,##0.0;[Red](#,##0.0);-";
  }
  if (col.includes("score") || col.includes("multiplier") || col.includes("zscore")) return "0.00";
  return "General";
}

function widthFor(columnName) {
  const col = columnName.toLowerCase();
  if (col.includes("date")) return 13;
  if (col.includes("issuer") || col.includes("bond_id") || col.includes("sector") || col.includes("regime")) return 18;
  if (col.includes("reason") || col.includes("scenario") || col.includes("parameter")) return 24;
  return 14;
}

function colToLetter(colNumberOneBased) {
  let dividend = colNumberOneBased;
  let columnName = "";
  while (dividend > 0) {
    const modulo = (dividend - 1) % 26;
    columnName = String.fromCharCode(65 + modulo) + columnName;
    dividend = Math.floor((dividend - modulo) / 26);
  }
  return columnName;
}

function toTableName(name) {
  return `${name.replace(/[^A-Za-z0-9]/g, "_")}_table`;
}

File: SystematicCredit/scripts/build_dashboard.py
#!/usr/bin/env python3
"""Build the Systematic Credit Excel dashboard from the pipeline payload."""

from __future__ import annotations

import json
import re
from datetime import datetime
from pathlib import Path
from typing import Any

import xlsxwriter


PROJECT_ROOT = Path(__file__).resolve().parents[1]
PAYLOAD_PATH = PROJECT_ROOT / "data" / "output" / "dashboard_payload.json"
OUTPUT_DIR = PROJECT_ROOT / "outputs" / "systematic_credit_toolkit"
OUTPUT_XLSX = OUTPUT_DIR / "bond_cds_basis_dashboard.xlsx"
WORKBOOK_ROW_LIMIT = 1_200

SHEET_ORDER = [
    "Dashboard",
    "CDS_Curves",
    "Hazard_Survival",
    "Bond_CDS_Basis",
    "RV_Signals",
    "Positions",
    "Macro_Regime",
    "Hedge_Overlay",
    "PnL_Attribution",
    "Performance",
    "Stress_Test",
    "Config",
]

TABLE_TO_SHEET = {
    "cds_curves": "CDS_Curves",
    "hazard_survival": "Hazard_Survival",
    "bond_cds_basis": "Bond_CDS_Basis",
    "rv_signals": "RV_Signals",
    "positions": "Positions",
    "macro": "Macro_Regime",
    "hedge_overlay": "Hedge_Overlay",
    "pnl_attribution": "PnL_Attribution",
    "performance": "Performance",
    "stress_test": "Stress_Test",
    "config": "Config",
}

COLORS = {
    "navy": "#12355B",
    "green": "#15803D",
    "gray": "#E5E7EB",
    "border": "#CBD5E1",
    "text": "#111827",
    "white": "#FFFFFF",
}


def main() -> None:
    if not PAYLOAD_PATH.exists():
        raise FileNotFoundError(
            f"Dashboard payload not found: {PAYLOAD_PATH}. "
            "Run scripts/run_pipeline.py first."
        )

    payload = json.loads(PAYLOAD_PATH.read_text(encoding="utf-8"))
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    workbook = xlsxwriter.Workbook(OUTPUT_XLSX)
    workbook.set_properties(
        {
            "title": "Bond-CDS Basis Trading Dashboard",
            "subject": "Systematic credit relative-value research",
            "author": "Q-Research",
        }
    )
    formats = build_formats(workbook)
    sheets = {name: workbook.add_worksheet(name) for name in SHEET_ORDER}
    for sheet in sheets.values():
        sheet.hide_gridlines(2)

    for table_name, sheet_name in TABLE_TO_SHEET.items():
        table = payload["tables"].get(table_name)
        if table:
            write_data_sheet(
                workbook,
                sheets[sheet_name],
                table_name,
                table,
                formats,
            )

    write_dashboard(workbook, sheets["Dashboard"], payload, formats)
    workbook.close()
    print(f"Dashboard written to {OUTPUT_XLSX}")


def build_formats(workbook: xlsxwriter.Workbook) -> dict[str, Any]:
    base = {"font_name": "Aptos", "font_color": COLORS["text"]}
    return {
        "title": workbook.add_format(
            {
                **base,
                "bold": True,
                "font_size": 16,
                "font_color": COLORS["white"],
                "bg_color": COLORS["navy"],
                "align": "left",
                "valign": "vcenter",
            }
        ),
        "subtitle": workbook.add_format(
            {
                **base,
                "italic": True,
                "bg_color": COLORS["gray"],
                "valign": "vcenter",
            }
        ),
        "header": workbook.add_format(
            {
                **base,
                "bold": True,
                "font_color": COLORS["white"],
                "bg_color": COLORS["navy"],
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "text": workbook.add_format({**base, "border": 1, "border_color": COLORS["border"]}),
        "currency": workbook.add_format(
            {
                **base,
                "num_format": "$#,##0;[Red]($#,##0);-",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "percent": workbook.add_format(
            {
                **base,
                "num_format": "0.0%;[Red](0.0%);-",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "number": workbook.add_format(
            {
                **base,
                "num_format": "0.00",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "integer": workbook.add_format(
            {
                **base,
                "num_format": "#,##0",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "bps": workbook.add_format(
            {
                **base,
                "num_format": "#,##0.0;[Red](#,##0.0);-",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
        "date": workbook.add_format(
            {
                **base,
                "num_format": "yyyy-mm-dd",
                "border": 1,
                "border_color": COLORS["border"],
            }
        ),
    }


def write_data_sheet(
    workbook: xlsxwriter.Workbook,
    sheet: xlsxwriter.worksheet.Worksheet,
    table_name: str,
    table: dict[str, Any],
    formats: dict[str, Any],
) -> None:
    columns = table["columns"]
    records = limit_workbook_rows(table["records"], table_name)

    for col_idx, column_name in enumerate(columns):
        sheet.write(0, col_idx, format_header(column_name), formats["header"])
        column_format = workbook.add_format(
            {
                "font_name": "Aptos",
                "font_color": COLORS["text"],
                "num_format": number_format_for(column_name),
                "border": 1,
                "border_color": COLORS["border"],
            }
        )
        sheet.set_column(col_idx, col_idx, width_for(column_name), column_format)

    for row_idx, record in enumerate(records, start=1):
        for col_idx, column_name in enumerate(columns):
            write_cell(sheet, row_idx, col_idx, record.get(column_name), column_name)

    last_row = max(len(records), 1)
    last_col = len(columns) - 1
    sheet.add_table(
        0,
        0,
        last_row,
        last_col,
        {
            "name": to_table_name(table_name),
            "style": "Table Style Medium 2",
            "columns": [{"header": format_header(column)} for column in columns],
        },
    )
    sheet.freeze_panes(1, 0)


def write_cell(
    sheet: xlsxwriter.worksheet.Worksheet,
    row: int,
    col: int,
    value: Any,
    column_name: str,
) -> None:
    if value is None:
        sheet.write_blank(row, col, None)
    elif "date" in column_name.lower() and isinstance(value, str):
        try:
            sheet.write_datetime(row, col, datetime.fromisoformat(value))
        except ValueError:
            sheet.write(row, col, value)
    else:
        sheet.write(row, col, value)


def limit_workbook_rows(records: list[dict[str, Any]], table_name: str) -> list[dict[str, Any]]:
    unlimited = {
        "macro",
        "hedge_overlay",
        "pnl_attribution",
        "performance",
        "stress_test",
        "config",
    }
    if table_name in unlimited or len(records) <= WORKBOOK_ROW_LIMIT:
        return records
    return records[-WORKBOOK_ROW_LIMIT:]


def write_dashboard(
    workbook: xlsxwriter.Workbook,
    sheet: xlsxwriter.worksheet.Worksheet,
    payload: dict[str, Any],
    formats: dict[str, Any],
) -> None:
    sheet.set_row(0, 24)
    sheet.merge_range("A1:H1", "Bond-CDS Basis Trading Dashboard", formats["title"])
    sheet.merge_range(
        "A2:H2",
        "Synthetic issuer-level research dashboard | CDS curves, basis signals, "
        "CS01 exposure, hedging, and P&L attribution",
        formats["subtitle"],
    )
    sheet.set_column("A:H", 15)
    sheet.set_column("I:Q", 12)

    write_kpis(sheet, payload, formats)
    helper_ranges = write_dashboard_helpers(sheet, payload, formats)
    add_dashboard_charts(workbook, sheet, helper_ranges)


def write_kpis(
    sheet: xlsxwriter.worksheet.Worksheet,
    payload: dict[str, Any],
    formats: dict[str, Any],
) -> None:
    performance = payload["tables"]["performance"]["records"][0]
    hedge_records = payload["tables"]["hedge_overlay"]["records"]
    latest_hedge = hedge_records[-1] if hedge_records else {}

    sheet.write_row("A4", ["Metric", "Value", "Format"], formats["header"])
    kpis = [
        ("Total P&L", performance.get("total_pnl"), "Currency", formats["currency"]),
        ("Total Return", performance.get("total_return"), "Percent", formats["percent"]),
        ("Sharpe Ratio", performance.get("sharpe_ratio"), "Number", formats["number"]),
        ("Max Drawdown", performance.get("max_drawdown"), "Percent", formats["percent"]),
        ("Hit Rate", performance.get("hit_rate"), "Percent", formats["percent"]),
        ("Ending NAV", performance.get("ending_nav"), "Currency", formats["currency"]),
    ]
    for row, (label, value, unit, value_format) in enumerate(kpis, start=4):
        sheet.write(row, 0, label, formats["text"])
        sheet.write(row, 1, value, value_format)
        sheet.write(row, 2, unit, formats["text"])

    sheet.write_row("E4", ["Latest Risk", "Value", "Unit"], formats["header"])
    risk_rows = [
        ("Macro Regime", latest_hedge.get("macro_regime"), "label", formats["text"]),
        ("Net Strategy CS01", latest_hedge.get("net_strategy_cs01"), "per bp", formats["integer"]),
        ("Gross Strategy CS01", latest_hedge.get("gross_strategy_cs01"), "per bp", formats["integer"]),
        ("Hedge Ratio", latest_hedge.get("hedge_ratio"), "%", formats["percent"]),
        ("Hedge CS01", latest_hedge.get("hedge_cs01"), "per bp", formats["integer"]),
        ("Hedge Notional", latest_hedge.get("hedge_notional"), "currency", formats["currency"]),
    ]
    for row, (label, value, unit, value_format) in enumerate(risk_rows, start=4):
        sheet.write(row, 4, label, formats["text"])
        sheet.write(row, 5, value, value_format)
        sheet.write(row, 6, unit, formats["text"])


def write_dashboard_helpers(
    sheet: xlsxwriter.worksheet.Worksheet,
    payload: dict[str, Any],
    formats: dict[str, Any],
) -> dict[str, tuple[int, int]]:
    pnl_records = payload["tables"]["pnl_attribution"]["records"][-60:]
    sheet.write_row("A13", ["Date", "Cumulative P&L"], formats["header"])
    for row, record in enumerate(pnl_records, start=13):
        date_value = record.get("date")
        try:
            date_value = datetime.fromisoformat(date_value) if isinstance(date_value, str) else date_value
        except ValueError:
            pass
        sheet.write(row, 0, date_value, formats["date"])
        sheet.write(row, 1, record.get("cumulative_total_pnl"), formats["currency"])

    summary = payload["tables"]["dashboard_summary"]["records"]
    latest_basis = [row for row in summary if row.get("summary_type") == "latest_basis"]
    sheet.write_row("D13", ["Bond", "Basis bps", "Z-score", "Signal"], formats["header"])
    for row, record in enumerate(latest_basis, start=13):
        sheet.write(row, 3, record.get("bond_id"), formats["text"])
        sheet.write(row, 4, record.get("bond_cds_basis_bps"), formats["bps"])
        sheet.write(row, 5, record.get("basis_zscore"), formats["number"])
        sheet.write(row, 6, record.get("rv_signal"), formats["text"])

    regime_counts = [row for row in summary if row.get("summary_type") == "regime_counts"]
    sheet.write_row("A77", ["Regime", "Days"], formats["header"])
    for row, record in enumerate(regime_counts, start=77):
        sheet.write(row, 0, record.get("macro_regime"), formats["text"])
        sheet.write(row, 1, record.get("days"), formats["integer"])

    attribution = [
        ("Basis Convergence", "basis_convergence_pnl", 1),
        ("Credit Spread", "credit_spread_pnl", 2),
        ("Hedge", "hedge_pnl", 5),
        ("Carry", "carry_pnl", 3),
        ("Transaction Costs", "transaction_costs", 4),
    ]
    sheet.write_row("D30", ["P&L Component", "Total"], formats["header"])
    all_pnl_records = payload["tables"]["pnl_attribution"]["records"]
    for row, (label, key, _) in enumerate(attribution, start=30):
        value = sum(float(record.get(key) or 0.0) for record in all_pnl_records)
        if key == "transaction_costs":
            value = -value
        sheet.write(row, 3, label, formats["text"])
        sheet.write(row, 4, value, formats["currency"])

    return {
        "pnl": (13, len(pnl_records)),
        "basis": (13, len(latest_basis)),
        "attribution": (30, len(attribution)),
        "regime": (77, len(regime_counts)),
    }


def add_dashboard_charts(
    workbook: xlsxwriter.Workbook,
    sheet: xlsxwriter.worksheet.Worksheet,
    ranges: dict[str, tuple[int, int]],
) -> None:
    pnl_start, pnl_count = ranges["pnl"]
    if pnl_count:
        chart = workbook.add_chart({"type": "line"})
        chart.add_series(
            {
                "name": "Cumulative P&L",
                "categories": ["Dashboard", pnl_start, 0, pnl_start + pnl_count - 1, 0],
                "values": ["Dashboard", pnl_start, 1, pnl_start + pnl_count - 1, 1],
                "line": {"color": COLORS["navy"], "width": 2.25},
            }
        )
        chart.set_title({"name": "Cumulative P&L"})
        chart.set_legend({"none": True})
        chart.set_y_axis({"num_format": "$#,##0"})
        sheet.insert_chart("J3", chart, {"x_scale": 1.0, "y_scale": 1.0})

    add_column_chart(workbook, sheet, "J20", "Latest Bond-CDS Basis (bps)", ranges["basis"], 3, 4, "#,##0.0")
    add_column_chart(workbook, sheet, "J37", "P&L Attribution", ranges["attribution"], 3, 4, "$#,##0")
    add_column_chart(workbook, sheet, "A88", "Macro Regime Days", ranges["regime"], 0, 1, "#,##0")


def add_column_chart(
    workbook: xlsxwriter.Workbook,
    sheet: xlsxwriter.worksheet.Worksheet,
    position: str,
    title: str,
    row_range: tuple[int, int],
    category_col: int,
    value_col: int,
    number_format: str,
) -> None:
    start, count = row_range
    if not count:
        return
    chart = workbook.add_chart({"type": "column"})
    chart.add_series(
        {
            "name": title,
            "categories": ["Dashboard", start, category_col, start + count - 1, category_col],
            "values": ["Dashboard", start, value_col, start + count - 1, value_col],
            "fill": {"color": COLORS["navy"]},
            "border": {"color": COLORS["navy"]},
        }
    )
    chart.set_title({"name": title})
    chart.set_legend({"none": True})
    chart.set_y_axis({"num_format": number_format})
    sheet.insert_chart(position, chart, {"x_scale": 1.0, "y_scale": 1.0})


def format_header(value: str) -> str:
    return str(value).replace("_", " ").title()


def number_format_for(column_name: str) -> str:
    column = column_name.lower()
    if "date" in column:
        return "yyyy-mm-dd"
    if any(token in column for token in ("return", "probability", "hazard_rate", "ratio", "drawdown", "hit_rate")):
        return "0.0%;[Red](0.0%);-"
    if any(token in column for token in ("pnl", "notional", "cost", "nav")):
        return "$#,##0;[Red]($#,##0);-"
    if any(token in column for token in ("bps", "cs01", "spread", "basis")):
        return "#,##0.0;[Red](#,##0.0);-"
    if any(token in column for token in ("score", "multiplier", "zscore")):
        return "0.00"
    return "General"


def width_for(column_name: str) -> int:
    column = column_name.lower()
    if "date" in column:
        return 13
    if any(token in column for token in ("issuer", "bond_id", "sector", "regime")):
        return 18
    if any(token in column for token in ("reason", "scenario", "parameter")):
        return 24
    return 14


def to_table_name(name: str) -> str:
    return f"{re.sub(r'[^A-Za-z0-9]', '_', name)}_table"


if __name__ == "__main__":
    main()
